include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -345,4 +350,164 @@ def ConcreteInputOp : VerifOp<"concrete_input", [
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// Formal Contract Ops
+//===----------------------------------------------------------------------===//
+
+class ContractLikeOp<string mnemonic, list<Trait> traits = []> 
+  : VerifOp<mnemonic, traits # [
+    HasParent<"hw::HWModuleOp">, RegionKindInterface, IsolatedFromAbove
+  ]> {
+
+  let arguments = (ins Variadic<AnyType>:$inputs);
+
+  let regions = (region SizedRegion<1>:$body);
+
+  let extraClassDeclaration = [{
+    /// Implement RegionKindInterface.
+    static RegionKind getRegionKind(unsigned index) { 
+      return RegionKind::Graph;
+    }
+
+    /// Retrieves the region block arguments
+    BlockArgument getRegionArg(unsigned index) {
+      return getBody().front().getArguments()[index];
+    }
+
+    /// Retrieves the number of block arguments
+    unsigned getNumRegionArgs() {
+      return getBody().front().getNumArguments();
+    }
+  }];
+
+  let hasRegionVerifier = 1;
+}
+
+
+def ContractOp : ContractLikeOp<"contract", [NoTerminator]>{
+  let summary = "declare a formal contract";
+  let description = [{
+    This operation declares a formal contract that is used to create precondition 
+    and postconditions on a parent module. These are used as an abstraction to 
+    better modularize formal verification such that each module containing a contract 
+    is checked exactly once.
+          
+    e.g. 
+    ```
+    hw.module @Bar(in %clk : !seq.clock, in %foo : i8, out "" : i8, out "1" : i8) { 
+      verif.contract (%clk, %foo) : (i8, i8) {
+        ^bb0(%clk : !seq.clock, %foo : i8): 
+          %c0_8 = hw.constant 0 : i8 
+          %prec = comb.icmp bin neq %c0_8, %foo : i8
+          verif.require %precond : i1
+
+          %out0, %out1 = verif.result : i8  
+          %post = comb.icmp bin ugt %out0, %foo : i8
+          %post1 = comb.icmp bin ult %out1, %foo : i8
+          verif.ensure %post : i1
+          verif.ensure %post1 : i1
+      }
+      /* ... Module definition ... */
+    }
+    ```
+
+    This later is used to replace any instance of Bar during verification:  
+    ```
+    %bar.0, %bar.1 = hw.instance "bar" @Bar("clk": %clk : !seq.clock, "foo" : %foo : i8)  -> ("" : i8, "1" : i8)
+
+    /* ... After PrepareForFormal Pass becomes ... */
+
+    %bar.0, %bar.1 = verif.instance (%clk, %foo) : (!seq.clock, i8) -> (i8, i8) {
+      ^bb0(%clk : !seq.clock, %foo : i8):
+        %c0_8 = hw.constant 0 : i8 
+        %prec = comb.icmp bin neq %c0_8, %foo : i8
+        verif.assert %precond : i1
+
+        %bar.0 = verif.symbolic_input : i8
+        %bar.1 = verif.symbolic_input : i8
+        %post = comb.icmp bin ugt %bar.0, %foo : i8
+        %post1 = comb.icmp bin ult %bar.1, %foo : i8
+        verif.assume %post : i1
+        verif.assume %post1 : i1
+        verif.yield %bar.0, %bar.1 : i8, i8
+    }
```

Comment:
I'm wondering if you need a `verif.instance` op at all. Could you instead just inline the contents of the contract? I think it only makes sense to add an operation if a pass is going to use that operation later. At first glance, `verif.instance` looks like it migth just be inlined during the formal checking of the contract. Or are there cases where you'd ignore the contents of `verif.instance`, or treat them differently than in other cases?

If the formal contract checking more or less inlines the contents of `verif.instance`, and it has to deal with things like `verif.symbolic_input` within the context of the surrounding module anyway, then you could simplify the IR by inlining everything directly and allowing `verif.symbolic_input` to be used anywhere in the module. WDYT?

Reply:
Given that symbolic inputs really only make sense in the context of verification, I really want to avoid having them be usable outside of explicit verification contexts. This will guarantee that they are always used legally in a way that doesn't require expensive verification logic. It is also a neat way to encapsulate the contract in a way that keeps the transformed IR similar to the base IR, thus making it easier to debug -- i.e. it is obvious that a verif.instance is the transformed result of the now missing hw.instance.